components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Runtime/Extensions/OpcConfigEx.cs

```diff
@@ -40,38 +41,58 @@ public static class OpcConfigEx {
 
             applicationConfiguration.CertificateValidator.CertificateValidation += handler;
 
-            X509Certificate2 certificate = null;
+            var configuredSubject = applicationConfiguration.SecurityConfiguration
+                .ApplicationCertificate.SubjectName;
+            applicationConfiguration.SecurityConfiguration.ApplicationCertificate.SubjectName = 
+                applicationConfiguration.ApplicationName;
+            applicationConfiguration.CertificateValidator
+                .Update(applicationConfiguration.SecurityConfiguration).ConfigureAwait(false);
 
-            // use existing certificate, if it is there
-            certificate = applicationConfiguration.SecurityConfiguration.ApplicationCertificate.Find(true).Result;
+            // use existing certificate, if present
+            var certificate = applicationConfiguration.SecurityConfiguration
+                .ApplicationCertificate.Certificate;
 
             // create a self signed certificate if there is none
             if (certificate == null && createSelfSignedCertIfNone) {
                 certificate = CertificateFactory.CreateCertificate(
-                    applicationConfiguration.SecurityConfiguration.ApplicationCertificate.StoreType,
-                    applicationConfiguration.SecurityConfiguration.ApplicationCertificate.StorePath,
+                    applicationConfiguration.SecurityConfiguration
+                        .ApplicationCertificate.StoreType,
+                    applicationConfiguration.SecurityConfiguration
+                        .ApplicationCertificate.StorePath,
                     null,
                     applicationConfiguration.ApplicationUri,
                     applicationConfiguration.ApplicationName,
-                    applicationConfiguration.ApplicationName,
+                    configuredSubject,
                     null,
                     CertificateFactory.defaultKeySize,
                     DateTime.UtcNow - TimeSpan.FromDays(1),
                     CertificateFactory.defaultLifeTime,
                     CertificateFactory.defaultHashSize
                 );
 
-                // update security information
-                applicationConfiguration.SecurityConfiguration.ApplicationCertificate.Certificate =
-                    certificate ??
+                if (certificate == null) {
                     throw new Exception(
                         "OPC UA application certificate can not be created! Cannot continue without it!");
-                //await applicationConfiguration.CertificateValidator.UpdateCertificate(
-                   //applicationConfiguration.SecurityConfiguration).ConfigureAwait(false);
+                }
+
+                applicationConfiguration.SecurityConfiguration
+                    .ApplicationCertificate.Certificate = certificate;
```

Comment:
>certificat [](start = 58, length = 10)

this is a cert with private key

Reply:
right the freshly created certificate has private key and is to be added to the own cert store